ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/PDEState.java

```diff
@@ -106,85 +105,61 @@ public int compare(BaseDescription o1, BaseDescription o2) {
 		}
 	}
 
-	private int compareDescription(BaseDescription bd1, BaseDescription bd2, String systemBSN) {
-		BundleDescription sup1 = bd1.getSupplier();
-		BundleDescription sup2 = bd2.getSupplier();
-		if (systemBSN.equals(sup1.getSymbolicName()) && !systemBSN.equals(sup2.getSymbolicName())) {
-			return -1;
-		} else if (!systemBSN.equals(sup1.getSymbolicName()) && systemBSN.equals(sup2.getSymbolicName())) {
-			return 1;
-		}
-		Version v1 = bd1.getVersion();
-		Version v2 = bd2.getVersion();
-		int versionCompare = versionCompare(v1, v2);
-		if (versionCompare != 0) {
-			return versionCompare;
-		}
-		BundleDescription s1 = bd1.getSupplier();
-		BundleDescription s2 = bd2.getSupplier();
-		// To have a stable ordering, the version MUST be compared
-		// first, even if this comes from different suppliers
-		int supplierVersionCompare = versionCompare(s1.getVersion(), s2.getVersion());
-		if (supplierVersionCompare != 0) {
-			return supplierVersionCompare;
-		}
-		String n1 = s1.getName();
-		String n2 = s2.getName();
-		if (n1 != null && n1.equals(n2)) {
-			boolean isQualifier = "qualifier".equals(v1.getQualifier()); //$NON-NLS-1$
-			if (!isQualifier) {
-				String loc1 = s1.getLocation();
-				String loc2 = s2.getLocation();
-				if (loc1 != null && loc2 != null && !loc1.equals(loc2)) {
-					IWorkspaceRoot root = ResourcesPlugin.getWorkspace().getRoot();
-					if (root != null) {
-						IPath p1 = new Path(loc1);
-						boolean p1InWorkspace = root.findContainersForLocationURI(URIUtil.toURI(p1)).length != 0;
-						IPath p2 = new Path(loc2);
-						boolean p2InWorkspace = root.findContainersForLocationURI(URIUtil.toURI(p2)).length != 0;
-						if (p1InWorkspace && !p2InWorkspace) {
-							return -1;
-						}
-						if (!p1InWorkspace && p2InWorkspace) {
-							return 1;
-						}
-					}
-				}
-			}
-		}
-		// as a last resort, compare the bundle id...
-		long id1 = s1.getBundleId();
-		long id2 = s2.getBundleId();
-		return (id1 < id2) ? -1 : ((id1 == id2) ? 0 : 1);
+	private Comparator<BaseDescription> createHigherLocalVersionFirstComparator(String systemBSN) {
+		return systemBundlesFirst(systemBSN) //
+				.thenComparing(BaseDescription::getVersion, HIGHER_VERSION) //
+				.thenComparing(BaseDescription::getSupplier,
```

Comment:
Seems  like now the supplier is (again) first compared, what should not be the case (see comment below)

Reply:
Just like before the suppliers are first compared by their version, then if they are in the workspace and last by their bundle-id.
But for me it looks like this is currently the case.

Reply:
Okay I was confused here maybe instead of using it inline extract the code below into a field like the other ones,

Reply:
OK, good. Did that.
Then the body of the method `createHigherLocalVersionFirstComparator()` is simple enough that it can be inlined into the only caller.